### PR TITLE
Stop instance fix (v0.0.11)

### PR DESCRIPTION
**Pull request by nivBlox**
This pr resolving 2 issues regarding the instance stopping (ghost) 

1. In case of force decided there was no preparedValue in instance state, what raised an error and stopped the single observer of stageChangeChan that caused the instance to be "stuck" on channel when trying to set stopped stage.
2. When instance is trying to stop timer but the timer already have been stop (didnt figure it out why stopped) the timer chan is being "drained" (observing the timer chan and wait for trigger)  but the timer will never call this chan again cause is consist single timer in instance

### CODE REVIEW

File: beacon/goclient/goclient.go [hunk 1]
```diff
@@ -75,13 +76,21 @@ func (gc *goClient) StartReceivingBlocks() {
 				gc.logger.Error("failed to fetch head block", zap.Error(err))
 				time.Sleep(time.Second * 2)
 				continue
+			} else if signedBeaconBlock == nil {
+				gc.logger.Debug("received empty block")
+				time.Sleep(time.Second * 1)
```
Comment (alonmuroch) [inline beacon/goclient/goclient.go:81]: @nivBlox why sleep? isn't it an error?

File: beacon/goclient/goclient.go [hunk 2]
```diff
@@ -53,6 +53,7 @@ func New(opt beacon.Options) (beacon.Beacon, error) {
 		network:  core.NetworkFromString(opt.Network),
 		client:   autoClient,
 		graffiti: []byte("BloxStaking"),
+		blockChannel: make(chan spec.SignedBeaconBlock),
```
Comment (alonmuroch) [inline beacon/goclient/goclient.go:56]: @nivBlox why did we add it? what was a before that?

File: config/config.yaml [hunk 3]
```diff
@@ -31,3 +31,5 @@ OperatorPrivateKey:
 bootnode:
   ExternalIP:
   PrivateKey:
+
```
Comment (alonmuroch) [inline config/config.yaml:34]: @nivBlox doesn't seem to be relevant to this PR

File: ibft/commit.go [hunk 4]
```diff
@@ -53,6 +53,9 @@ func (i *Instance) forceDecidedPipeline() pipeline.Pipeline {
 
 // CommittedAggregatedMsg returns a signed message for the state's committed value with the max known signatures
 func (i *Instance) CommittedAggregatedMsg() (*proto.SignedMessage, error) {
+	if i.State.DecidedMsg != nil {
+		return i.State.DecidedMsg, nil
+	}
 	if i.State.PreparedValue == nil {
```
Comment (alonmuroch) [inline ibft/commit.go:59]: @nivBlox why do we need anything after i.State.DecidedMsg? if it's not decided than i.State.DecidedMsg is nil and so does the code after it

File: ibft/commit.go [hunk 5]
```diff
@@ -99,7 +102,21 @@ func (i *Instance) uponCommitMsg() pipeline.Pipeline {
 			i.Logger.Info("commit iBFT instance",
 				zap.String("Lambda", hex.EncodeToString(i.State.Lambda)), zap.Uint64("round", i.State.Round),
 				zap.Int("got_votes", len(sigs)))
-
+			var decided *proto.SignedMessage
```
Comment (alonmuroch) [inline ibft/commit.go:105]: @nivBlox I think we should have a function to aggregate separate sigs, please refactor

File: ibft/event_loop.go [hunk 6]
```diff
@@ -54,14 +54,20 @@ loop:
 
 		var wg sync.WaitGroup
 		if i.MsgQueue.MsgCount(msgqueue.IBFTMessageIndexKey(i.State.Lambda, i.State.SeqNumber, i.State.Round)) > 0 {
+			i.Logger.Debug("adding ibft message to event queue - waiting for done")
 			wg.Add(1)
-			i.eventQueue.Add(func() {
+			if added := i.eventQueue.Add(func() {
```
Comment (alonmuroch) [inline ibft/event_loop.go:59]: @nivBlox I don't think we should return bool from add... what's the rational?

File: ibft/eventqueue/queue.go [hunk 7]
```diff
@@ -1,36 +1,54 @@
...
+	"time"
+)
 
-// Queue is a simple struct that manages a queue of functions. Thread safe
-type Queue struct {
+// Event represent some function
+type Event func()
+
+// EventQueue is the interface for managing a queue of functions
+type EventQueue interface {
```
Comment (alonmuroch) [inline ibft/eventqueue/queue.go:14]: @nivBlox why do we need Ann interface? Even if we do it should be in a different PR

File: ibft/eventqueue/queue.go [hunk 8]
```diff
@@ -1,36 +1,54 @@
...
+	q := queue{
+		queue: make([]Event, 0),
 		lock:  sync.Mutex{},
 	}
+	return &q
 }
 
 // Add will add an an item to the queue, thread safe.
-func (q *Queue) Add(f func()) {
+func (q *queue) Add(e Event) bool {
```
Comment (alonmuroch) [inline ibft/eventqueue/queue.go:37]: @nivBlox why return bool from add?

File: ibft/ibft_decided_test.go [hunk 9]
```diff
@@ -251,6 +254,52 @@ func TestDecideIsCurrentInstance(t *testing.T) {
 	}
 }
 
+func TestForceDecided(t *testing.T) {
```
Comment (alonmuroch) [inline ibft/ibft_decided_test.go:257]: @nivBlox this doesn't test force decided but rather sync

File: ibft/ibft_network.go [hunk 10]
```diff
@@ -38,6 +38,9 @@ func (i *ibftImpl) listenToNetworkMessages() {
 	go func() {
 		for msg := range msgChan {
 			if msg.Message != nil && i.equalIdentifier(msg.Message.Lambda) {
+				i.logger.Debug("adding ibft msg to queue",
```
Comment (alonmuroch) [inline ibft/ibft_network.go:41]: @nivBlox do we need this?

File: ibft/ibft_network.go [hunk 11]
```diff
@@ -52,6 +55,8 @@ func (i *ibftImpl) listenToNetworkDecidedMessages() {
 	go func() {
 		for msg := range decidedChan {
 			if msg.Message != nil && i.equalIdentifier(msg.Message.Lambda) {
+				i.logger.Debug("adding decided msg to queue",
```
Comment (alonmuroch) [inline ibft/ibft_network.go:58]: @nivBlox do we need this?

File: ibft/ibft_network.go [hunk 12]
```diff
@@ -67,6 +72,8 @@ func (i *ibftImpl) listenToSyncMessages() {
 	go func() {
 		for msg := range syncChan {
 			if msg.Msg != nil && i.equalIdentifier(msg.Msg.Lambda) {
+				i.logger.Debug("adding sync msg to queue",
```
Comment (alonmuroch) [inline ibft/ibft_network.go:75]: @nivBlox do we need this?

File: ibft/ibft_running_instance.go [hunk 13]
```diff
@@ -48,7 +48,9 @@ instanceLoop:
 		}
 	}
 	// when main instance loop breaks, nil current instance
+	i.logger.Debug("instance existed, nilling currentInstance", zap.Error(err))
```
Comment (alonmuroch) [inline ibft/ibft_running_instance.go:51]: @nivBlox remove

File: ibft/ibft_running_instance.go [hunk 14]
```diff
@@ -48,7 +48,9 @@ instanceLoop:
...
 	i.currentInstance = nil
+	i.logger.Debug("instance nilled")
```
Comment (alonmuroch) [inline ibft/ibft_running_instance.go:53]: @nivBlox let's change to iBFT instance result looop stopped

File: ibft/instance.go [hunk 15]
```diff
@@ -175,33 +175,44 @@ func (i *Instance) ForceDecide(msg *proto.SignedMessage) {
...
-			i.stageLock.Lock()
-			defer i.stageLock.Unlock()
-			if i.stageChangedChan != nil {
-				close(i.stageChangedChan)
-				i.stageChangedChan = nil
-			}
-
-			i.Logger.Info("stopped iBFT instance")
-		})
+		if added := i.eventQueue.Add(i.stop); !added {
```
Comment (alonmuroch) [inline ibft/instance.go:178]: @nivBlox I like it less than run once...

File: ibft/instance.go [hunk 16]
```diff
@@ -175,33 +175,44 @@ func (i *Instance) ForceDecide(msg *proto.SignedMessage) {
...
+			i.Logger.Debug("could not add 'stop' to event queue")
+		}
 	})
 }
 
-// Stopped returns true if instance is stopped
+// stop stops the instance
+func (i *Instance) stop() {
```
Comment (alonmuroch) [inline ibft/instance.go:185]: @nivBlox I don't really get why we should refactor?

File: ibft/msg_processing.go [hunk 17]
```diff
@@ -11,6 +11,9 @@ import (
 // ProcessMessage pulls messages from the queue to be processed sequentially
 func (i *Instance) ProcessMessage() (processedMsg bool, err error) {
 	if netMsg := i.MsgQueue.PopMessage(msgqueue.IBFTMessageIndexKey(i.State.Lambda, i.State.SeqNumber, i.State.Round)); netMsg != nil {
+		i.Logger.Debug("ibft message was popped from queue",
```
Comment (alonmuroch) [inline ibft/msg_processing.go:14]: @nivBlox remove

File: ibft/msg_processing.go [hunk 18]
```diff
@@ -30,6 +33,7 @@ func (i *Instance) ProcessMessage() (processedMsg bool, err error) {
 		}
 		return true, nil
 	}
+	i.Logger.Debug("could not find ibft message in queue")
```
Comment (alonmuroch) [inline ibft/msg_processing.go:36]: @nivBlox remove

File: ibft/roundtimer/timer.go [hunk 19]
```diff
@@ -60,19 +61,20 @@ func (t *RoundTimer) Stopped() bool {
...
+			t.cancelC <- true
 		}
-		t.cancelC <- true
 	}
 }
 
 func (t *RoundTimer) eventLoop() {
 	select {
 	case <-t.timer.C:
+		logex.GetLogger().Debug("eventLoop timerChan cancel")
```
Comment (alonmuroch) [inline ibft/roundtimer/timer.go:73]: @nivBlox remove

File: ibft/roundtimer/timer.go [hunk 20]
```diff
@@ -60,19 +61,20 @@ func (t *RoundTimer) Stopped() bool {
...
 		t.resC <- true
 		t.markStopped()
 	case <-t.cancelC:
+		logex.GetLogger().Debug("eventLoop cancelChan trigger")
```
Comment (alonmuroch) [inline ibft/roundtimer/timer.go:77]: @nivBlox remove

File: metrics/handler.go [hunk 21]
```diff
@@ -32,6 +33,9 @@ type metricsHandler struct {
 func (mh *metricsHandler) Start(mux *http.ServeMux, addr string) error {
 	mh.logger.Info("setup metrics collection")
 
+	runtime.SetBlockProfileRate(1000)
```
Comment (alonmuroch) [inline metrics/handler.go:36]: @nivBlox not sure it's relevant.. if it does it should be In a different PR

File: network/msgqueue/message_queue.go [hunk 22]
```diff
@@ -20,18 +24,22 @@ type messageContainer struct {
 // To solve this issue we have a message broker from which the instance pulls new messages, this also reduces concurrency issues as the instance is now single threaded.
 // The message queue has internal logic to organize messages by their round.
 type MessageQueue struct {
-	msgMutex    sync.Mutex
+	msgMutex    sync.RWMutex
 	indexFuncs  []IndexFunc
-	queue       map[string][]*messageContainer // = map[index][messageContainer.id]messageContainer
-	allMessages map[string]*messageContainer
+	queue       map[string][]messageContainer // = map[index][messageContainer.id]messageContainer
```
Comment (alonmuroch) [inline network/msgqueue/message_queue.go:29]: @nivBlox not sure this is relevant to this PR.. might be a good idea to separate

File: validator/metrics/collector.go [hunk 23]
```diff
@@ -85,6 +86,18 @@ func (c *validatorsCollector) Collect() ([]string, error) {
 				runningIbftsCountValidator, hex.EncodeToString(pk), runningIbftsValidator))
 		}
 
+		if v, exist := c.validatorCtrl.GetValidator(pubKey.SerializeToHexStr()); exist {
```
Comment (alonmuroch) [inline validator/metrics/collector.go:89]: @nivBlox not sure it's relevant, might be better to separate to a different PR

File: validator/validator.go [hunk 24]
```diff
@@ -167,3 +167,9 @@ func (v *Validator) GetIBFT(r beacon.RoleType) (ibft.IBFT, bool) {
 	i, exist := v.ibfts[r]
 	return i, exist
 }
+
+// GetQueue returns msgQueue
+func (v *Validator) GetQueue() *msgqueue.MessageQueue {
```
Comment (alonmuroch) [inline validator/validator.go:172]: @nivBlox why do we need this?

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: beacon/goclient/goclient.go (hunk 2 above)
Comment (nivBlox) [reply beacon/goclient/goclient.go:56]: yes it was but for now we disabled it as discussed

File: config/config.yaml (hunk 3 above)
Comment (nivBlox) [reply config/config.yaml:34]: its pretty code

File: ibft/roundtimer/timer.go (hunk 20 above)
Comment (nivBlox) [reply ibft/roundtimer/timer.go]: done

File: metrics/handler.go (hunk 21 above)
Comment (nivBlox) [reply metrics/handler.go:36]: @amir-blox is moving it to differ pr

File: network/msgqueue/message_queue.go (hunk 22 above)
Comment (nivBlox) [reply network/msgqueue/message_queue.go:29]: @amir-blox is cleaning this area

File: validator/metrics/collector.go (hunk 23 above)
Comment (nivBlox) [reply validator/metrics/collector.go:89]: @amir-blox is on it

File: validator/validator.go (hunk 24 above)
Comment (nivBlox) [reply validator/validator.go:172]: @amir-blox related to metric (moving to separate pr)

File: ibft/commit.go (hunk 4 above)
Comment (nivBlox) [reply ibft/commit.go]: done

File: ibft/commit.go (hunk 5 above)
Comment (nivBlox) [reply ibft/commit.go]: done

File: beacon/goclient/goclient.go (hunk 1 above)
Comment (nivBlox) [reply beacon/goclient/goclient.go:81]: clients in some cases returning empty block. no need special handling

File: metrics/handler.go (hunk 21 above)
Comment (amir-blox) [reply metrics/handler.go:36]: Added in https://github.com/bloxapp/ssv/pull/215

File: network/msgqueue/message_queue.go (hunk 22 above)
Comment (amir-blox) [reply network/msgqueue/message_queue.go:29]: added in https://github.com/bloxapp/ssv/pull/214

File: beacon/goclient/goclient.go (hunk 1 above)
Comment (alonmuroch) [reply beacon/goclient/goclient.go:81]: @nivBlox but why sleep?

File: beacon/goclient/goclient.go (hunk 2 above)
Comment (alonmuroch) [reply beacon/goclient/goclient.go:56]: @nivBlox so should we remove it?

File: validator/validator.go (hunk 24 above)
Comment (alonmuroch) [reply validator/validator.go:172]: @nivBlox doesn't seem to be moved

File: beacon/goclient/goclient.go [hunk 25]
```diff
@@ -111,9 +120,9 @@ func (gc *goClient) GetIndices(validatorPubKeys []spec.BLSPubKey) (map[spec.Vali
 //   (b) one-third of the slot has transpired (SECONDS_PER_SLOT / 3 seconds after the start of slot)
 func (gc *goClient) waitOneThirdOrValidBlock(slot uint64) {
 	// Don't need to wait if requested slot is the same as highest valid slot.
-	if slot <= gc.highestValidSlot {
-		return
-	}
+	//if slot <= gc.highestValidSlot {
```
Comment (alonmuroch) [inline beacon/goclient/goclient.go:123]: @nivBlox cleanup if not needed

File: ibft/change_round.go [hunk 26]
```diff
@@ -47,11 +47,14 @@ upon receiving a quorum Qrc of valid ⟨ROUND-CHANGE, λi, ri, −, −⟩ messa
 */
 func (i *Instance) uponChangeRoundFullQuorum() pipeline.Pipeline {
 	return pipeline.WrapFunc("upon change round full quorum", func(signedMessage *proto.SignedMessage) error {
+		i.Logger.Info("received valid change round message",
```
Comment (alonmuroch) [inline ibft/change_round.go:50]: @nivBlox let's remove it

File: ibft/commit.go [hunk 27]
```diff
@@ -100,6 +76,7 @@ func (i *Instance) uponCommitMsg() pipeline.Pipeline {
 				zap.String("Lambda", hex.EncodeToString(i.State.Lambda)), zap.Uint64("round", i.State.Round),
 				zap.Int("got_votes", len(sigs)))
 
+			i.State.DecidedMsg = i.aggregateMessages(sigs)
```
Comment (alonmuroch) [inline ibft/commit.go:79]: @nivBlox shouldn't we check if i.aggregateMessages returns nil or not?

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: ibft/eventqueue/queue.go (hunk 8 above)
Comment (amir-blox) [reply ibft/eventqueue/queue.go:34]: @alonmuroch to identify cases where the event wasn't added to the queue (if it was stopped)

File: beacon/goclient/goclient.go (hunk 1 above)
Comment (amir-blox) [reply beacon/goclient/goclient.go:81]: @alonmuroch to act similarly to error and success cases, both have time.sleep

File: beacon/goclient/goclient.go (hunk 2 above)
Comment (nivBlox) [reply beacon/goclient/goclient.go:56]: removed

File: beacon/goclient/goclient.go (hunk 25 above)
Comment (nivBlox) [reply beacon/goclient/goclient.go]: removed

File: ibft/event_loop.go (hunk 6 above)
Comment (nivBlox) [reply ibft/event_loop.go:60]: we using wait group for those functions and in case the msg is not added (cause of queue stopped) we have blocked function

File: ibft/eventqueue/queue.go (hunk 7 above)
Comment (nivBlox) [reply ibft/eventqueue/queue.go:11]: why in differ pr?

File: ibft/ibft_decided_test.go (hunk 9 above)
Comment (nivBlox) [reply ibft/ibft_decided_test.go:257]: why?

File: ibft/instance.go (hunk 15 above)
Comment (nivBlox) [reply ibft/instance.go:182]: its not instead. just making sure the msg is added properly

File: ibft/instance.go (hunk 16 above)
Comment (nivBlox) [reply ibft/instance.go:189]: its not really refactoring but rather extract to separate func

**Review by lior-blox — APPROVED** [review]
